### PR TITLE
Migrate std::complex to c10::complex

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Hope this fixes the xla build failure at https://github.com/pytorch/pytorch/pull/39838

### CODE REVIEW

File: torch_xla/csrc/tensor_util.cpp [hunk 1]
```diff
@@ -130,16 +130,16 @@ struct Caster<xla::half> {
   }
 };
 template <>
-struct Caster<std::complex<float>> {
```
Comment (dlibenzi) [inline torch_xla/csrc/tensor_util.cpp:133]: Cannot replace, as this is also used to cast XLA data, which uses `std::complex`.

Comment (zasdfgbnm) [reply torch_xla/csrc/tensor_util.cpp:133]: Hi @dlibenzi, thanks for the hint! I am a bit confused by how this Caster work: what would happen if the casting is from `std::complex<double>` to `std::complex<float>`? Will it discard the imag part of the number?

**Comment by zasdfgbnm** [comment]
sure

**Comment by zasdfgbnm** [comment]
squashed

**Comment by zasdfgbnm** [comment]
I am not sure what I am currently doing is the best approach. If XLA internally is using `std::complex` while PyTorch is using `c10::complex`, there has to be a casting somewhere. For better engineering, it might be good to keep the code consistent, instead of mixing different complexes everywhere. So, maybe a better solution is just something like below?
```C++
    case at::ScalarType::ComplexFloat:
      return xla::util::DataHash(reinterpret_cast<std::complex<float>*>(ctensor.data_ptr<c10::complex<float>>()), size);
    case at::ScalarType::ComplexDouble:
      return xla::util::DataHash(reinterpret_cast<std::complex<double>*>(ctensor.data_ptr<c10::complex<double>>()),
                                 size);
```
?

**Comment by zasdfgbnm** [comment]
Not sure if the test failure is related or not.

**Comment by dlibenzi** [comment]
We know that error but let me re-run the tests ...

**Pull request merged**